src/Workspaces/Core/Portable/FlowAnalysis/CustomDataFlowAnalysis.cs

```diff
@@ -0,0 +1,394 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.FlowAnalysis
+{
+    internal static class CustomDataFlowAnalysis<TBlockAnalysisData>
```

Comment:
given that it is cloned, we likely want to put comments on both sides mentioning this.  If one side changes, we likely will want to keep the other mostly in sync.  It will also avoid requests to change things that will cause things to diverge.

Reply:
> with a few modifications to track custom data instead of just a bool

could you dive deeper in this?  what sort of custom data are you tracking?

Reply:
Thanks.

---
In reply to: [231665789](https://github.com/dotnet/roslyn/pull/31025#discussion_r231665789) [](ancestors = 231665789)

Reply:
Also, is there a way that an even more common base type than this could be extracted, which jsut exposes helper abstract methods that operate with the 'custom data' type each side wants to track?  compiler would instantiate with 'bool' and have the helps that manage that.  you could have your own type.  Just a thought :)

Reply:
This could be any custom data. For example, for SymbolUsageAnalysis being added here, we track this BasicBlockAnalysisData data per basic block: [https://github.com/dotnet/roslyn/pull/31025/files#diff-4b7acefa197b3d57508ea48c29c892c4R19](https://github.com/dotnet/roslyn/pull/31025/files#diff-4b7acefa197b3d57508ea48c29c892c4R19)

---
In reply to: [231666116](https://github.com/dotnet/roslyn/pull/31025#discussion_r231666116) [](ancestors = 231666116)

Reply:
Yes, certainly. Once we have more analyses based on this framework, we can extract out a common base type. Currently, the only product implementation is in SymbolUsageAnalysis added here. The 'bool' based compiler one is a test only implementation.

---
In reply to: [231666807](https://github.com/dotnet/roslyn/pull/31025#discussion_r231666807) [](ancestors = 231666807)

Reply:
i thought that might be helpful as a cleaner design since it would help separate out logic that one really shouldn't be changing, versus the extension points they would be expected to implement.  but that could certainly be done in the future as well.